design/ceph/osd-migration.md

```diff
@@ -0,0 +1,239 @@
+- After upgrading an existing rook ceph cluster to a Ceph version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  osdStore:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### API
+
+```go
+// OSDReplaceInfo represents an OSD that needs to replaced
+type OSDReplaceInfo struct {
+	ID   int    `json:"id"`
+	Path string `json:"path"`
+	Node string `json:"node"`
+}
+
+// ReplaceQueue contains the list of OSDs to be replaced one at a time
+type ReplaceQueue struct {
+	OSDs []OSDReplaceInfo
+}
+
+// OSDReplaceConfig contains the status of the OSD replacement
+type OSDReplaceConfig struct {
+	OSDs                *ReplaceQueue    `json:"osds"`
+	IsPgHealthy          bool            `json:"isPgHealthy"`
+	PgHealthMessage      string          `json:"pgHealthMessage"`
+  OSDInProgress        *OSDReplaceInfo `json:"OSDInProgress"`
+	LastReplacedOSD      *OSDReplaceInfo `json:"lastReplacedOSD"`
+	LastReplacedTimeStamp time.Time      `json:"lastReplacedTimeStamp"`
+	StartTimeStamp       time.Time       `json:"startTimeStamp"`
+}
+```
+
+**OSDReplaceInfo**: Represents an OSD that needs to be replaced
+  - **ID**: OSD ID of the OSD to be replaced.
+  - **Path**: Disk or pvc claim that the OSD is consuming.
+  - **Node**: Node on which the OSD is running
+
+
+**ReplaceQueue**: This is a queue implementation containing a list of OSDs to be replaced. It will support basic queue operations like push, pop, size etc.
+
+
+**OSDReplaceConfig**: This will hold the current status/progress of the OSD replacement in the cluster.
+  - **OSDs**:  List of OSDs that are yet to be replaced.
+  - **IsPgHealth**: Whether Placement Groups are healthy or not. This is a boolean field.
+  - **PgHealthMessage**: This will contain a detailed health message of the placement groups. For example: `active+clean`
+  - **OSDInProgress**: Respresents the OSDs for which the deployment is being deleted.
+  - **LastReplacedOSD**: This field will contain details of the OSD that was last replaced.
+  - **LastReplacedTimeStamp**: Time at which the last OSD was replaced.
+  - **StartTimeStamp**: Time at which the whole OSD replacement process was started. This will help users know how much time has elapsed since the first OSD was replaced.
+
+
+#### Replace OSD :
+
+Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label not matching the osd store type provided in `spec.storage.osdStore.type`. This list will be stored in `OSDReplaceConfig.OSDs` and will be used a queue.
+  2. If the PG status is not `active+clean`, don't replace any OSD.
```

Comment:
From #12612 we likely need a common helper for determining health status from PGs, if it is safe to proceed with OSDs.

Reply:
we have a [helper](https://github.com/rook/rook/blob/52c9cfea9fe3c7e30f044040d468399f008379d5/pkg/daemon/ceph/client/status.go#L220) to determine this